### PR TITLE
🐛 Source Stripe: adding integration tests

### PR DESCRIPTION
**Pull request by maxi297**
## What
Adding integration tests for source-stripe events stream

## How
* Add some tooling as part of https://github.com/airbytehq/airbyte/pull/33305
* Temporarily upgrade the CDK version to fix breaking change (will need to upgrade again to have access to the changes mentioned above)
* Ensuring backoff is very small to have fast running tests
* Create a ConfigBuilder to ease testing
* Add the tests

## Recommended reading order
1. `airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py``
2. The rest

### CODE REVIEW

**Comment by maxi297** [comment]
Currently failing because some of the tooling is not release until [this](https://github.com/airbytehq/airbyte/pull/33305) is released

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py [hunk 1]
```diff
@@ -21,6 +21,7 @@
 
 STRIPE_API_VERSION = "2022-11-15"
 CACHE_DISABLED = os.environ.get("CACHE_DISABLED")
+IS_TESTING = os.environ.get("IS_TESTING")
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:24]: How can we detect if it's OSS or cloud? Those variables are only set during tests for now

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/config.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+from datetime import datetime
+from typing import Any, Dict
+
+
+class ConfigBuilder:
+    def __init__(self) -> None:
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/config.py:6]: I don't mind using dicts. I mind when they are exposed as this is not a domain object and is susceptible to change. In this case, it is the internals of the class and the blast radius is quite small if there is a change.

Now, I'm saying this and `build` returns a `Dict[str, Any]`... We don't really have a domain object for this. We could type it as `TConfig` though but it does not change anything as this change is only esthetic

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+) -> EntrypointOutput:
+    catalog = _catalog(sync_mode)
+    return read(_source(catalog), config.build(), catalog, state, expecting_exception)
+
+
+@freezegun.freeze_time(_NOW.isoformat())
+class FullRefreshTest(TestCase):
+
+    @HttpMocker()
+    def test_given_one_page_when_read_then_return_record(self, http_mocker: HttpMocker) -> None:
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:82]: I can add another record to this one and change the assert, yes

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+                },
+                headers=_AUTHENTICATION_HEADERS,
+            ),
+            _a_response().with_record(_a_record()).build()
+        )
+        output = self._read(_config().with_start_date(_A_START_DATE))
+        assert len(output.records) == 2
+
+    @HttpMocker()
+    def test_given_start_date_before_30_days_stripe_limit_when_read_then_request_more_than_30_days(self, http_mocker: HttpMocker) -> None:
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:121]: It is a special case because Stripe will only ever return 30 days worth of data on events. The [doc](https://stripe.com/docs/api/events) mentions: `We only guarantee access to events through the [Retrieve Event API](https://stripe.com/docs/api/events#retrieve_event) for 30 days.` Hence, this feels noteworthy

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+            ),
+            _a_response().build()
+        )
+
+        self._read(_config().with_start_date(start_date).with_lookback_window_in_days(lookback_window.days))
+
+        # request matched http_mocker
+
+    @HttpMocker()
+    def test_given_slice_range_when_read_then_request_before_start_date(self, http_mocker: HttpMocker) -> None:
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:152]: You are right! I updated it

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+                url="https://api.stripe.com/v1/events",
+                query_params=ANY_QUERY_PARAMS,
+                headers=_AUTHENTICATION_HEADERS
+            ),
+            HttpResponse(
+                json.dumps({
+                    "error": {
+                        "code": "parameter_invalid_integer",
+                        "doc_url": "https://stripe.com/docs/error-codes/parameter-invalid-integer",
+                        "message": "Invalid integer: not_an_integer",
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:188]: I could not find a case in prod so I created mine. Do you have a specific case in mind or a way to find one?

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+                        "type": "invalid_request_error"
+                    }
+                }),
+                400)
+        )
+        output = self._read(_config().with_start_date(_A_START_DATE))
+        assert len(output.get_stream_statuses(_STREAM_NAME)) == 0
+
+    @HttpMocker()
+    def test_given_http_status_401_when_read_then_stream_is_incomplete(self, http_mocker: HttpMocker) -> None:
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:200]: 401 (i.e. authentication issues) are still a prevalent cases we face in general for our connectors. I don't know if it's the case per se for stripe. I would assume not because right now, the failure type is `system_error` and not `config_error`. 

My intent was to show that we would be paged given users are unable to log in. Because of [this code block](https://github.com/airbytehq/airbyte/blob/7a9ee9adec0994769f25b6ec24146a5375c41b43/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py#L135-L140), I assumed the status INCOMPLETE would be sufficient but checking this again, it feels it does not mean the stream will break. Hence, the assertion should be updated to `output.errors[0].trace.error.failure_type`. Does that make sense?

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+
+    @HttpMocker()
+    def test_when_read_then_validate_availability_for_full_refresh_and_incremental(self, http_mocker: HttpMocker) -> None:
+        request = HttpRequest(url="https://api.stripe.com/v1/events", query_params=ANY_QUERY_PARAMS, headers=_AUTHENTICATION_HEADERS)
+        http_mocker.get(
+            request,
+            _a_response().build()
+        )
+        self._read(_config().with_start_date(_A_START_DATE))
+        http_mocker.assert_number_of_calls(request, 3)  # one call for full_refresh availability, one call for incremental availability and one call for the actual read
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:280]: Not a CDK bug but a stripe one if it's one. See https://github.com/airbytehq/airbyte/blob/7a9ee9adec0994769f25b6ec24146a5375c41b43/airbyte-integrations/connectors/source-stripe/source_stripe/availability_strategy.py#L75-L78

Basically, as long as the stream is available in full_refresh and supports incremental, we check availability for both

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+                query_params={"created[gte]": str(int(_A_START_DATE.timestamp())), "created[lte]": str(int(_NOW.timestamp())), "limit": 100},
+                headers=_AUTHENTICATION_HEADERS
+            ),
+            _a_response().with_record(_a_record().with_cursor(cursor_value)).build()
+        )
+        output = self._read(_config().with_start_date(_A_START_DATE), {})
+        assert output.most_recent_state == {"events": {"created": cursor_value}}
+
+    @HttpMocker()
+    def test_given_state_when_read_then_use_state_for_query_params(self, http_mocker: HttpMocker) -> None:
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:304]: The stream slices is a implementation detail. At this level, the input that affect the HTTP requests is the state

File: airbyte-integrations/connectors/source-stripe/unit_tests/resource/http/response/events.json [hunk 10]
```diff
@@ -0,0 +1,58 @@
+{
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/resource/http/response/events.json:1]: It is used. [This](https://github.com/airbytehq/airbyte/pull/33306/files#diff-627d3d51c032999499b527e030d7b5e6423f9bfdef3aa0ea9869ed4569e00205R52) is where the magic happens

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+
+def _catalog(sync_mode: SyncMode) -> ConfiguredAirbyteCatalog:
+    return CatalogBuilder().with_stream(_STREAM_NAME, sync_mode).build()
+
+
+def _source(catalog: ConfiguredAirbyteCatalog) -> SourceStripe:
+    return SourceStripe(catalog)
+
+
+def _create_builders() -> Tuple[RecordBuilder, HttpResponseBuilder]:
```
Comment (clnoll) [inline airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:51]: Could we revisit some of the function names and signatures? For example, I'm not finding it obvious that something called `create_builders_from_resource` or `_create_builders` would return this specific Tuple, and it's causing us to have to do some indexing in `_a_record` and `_a_response` that I don't love. I could understand keeping them coupled if we were using the record & response together, but looking at the usages it doesn't seem like we are.

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+
+
+@freezegun.freeze_time(_NOW.isoformat())
+class FullRefreshTest(TestCase):
+
+    @HttpMocker()
+    def test_given_one_page_when_read_then_return_record(self, http_mocker: HttpMocker) -> None:
+        http_mocker.get(
+            HttpRequest(
+                url="https://api.stripe.com/v1/events",
```
Comment (clnoll) [inline airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:86]: Are you thinking of this test suite as testing the Events stream or the `https://api.stripe.com/v1/events` endpoint? I think the distinction there is that if we're testing the stream, we may want to package up things like the url and query params in some kind of representative object to access them here; if the definition of "Events stream" changes we'll just have to change the one place. If the latter, the url can probably just be a constant, but we may want to be explicit about what query params are expected/required somewhere.

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+                },
+                headers=_AUTHENTICATION_HEADERS,
+            ),
+            _a_response().with_record(_a_record()).with_record(_a_record()).build()
+        )
+        output = self._read(_config().with_start_date(_A_START_DATE))
+        assert len(output.records) == 3
+
+    @HttpMocker()
+    def test_given_start_date_before_30_days_stripe_limit_and_slice_range_when_read_then_perform_request_before_30_days(self, http_mocker: HttpMocker) -> None:
```
Comment (clnoll) [inline airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:122]: For test cases like this that are tied specifically to Stripe, feels like a good idea to link to documentation and possibly provide a comment with additional context (e.g. in case the documentation gets stale). I also wonder if we should organize them differently than tests that cover behavior that we'd expect all sources to follow. I think it would be awesome of we could make the integration test suite a sort of living document that highlights caveats/interesting aspects of a specific source, so that people who are maintaining or debugging the connector can get a sense of what makes this connector different.

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+                    }
+                }),
+                401
+            )
+        )
+        output = self._read(_config().with_start_date(_A_START_DATE), expecting_exception=True)
+        assert output.get_stream_statuses(_STREAM_NAME) == [AirbyteStreamStatus.INCOMPLETE]
+
+    @HttpMocker()
+    def test_given_rate_limited_when_read_then_retry_and_return_records(self, http_mocker: HttpMocker) -> None:
```
Comment (clnoll) [inline airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:262]: WDYT about having a wrapper around `unittest.TestCase` (e.g. `BaseFullRefreshTest` and `BaseIncrementalTest`) that provides abstractmethods for test cases that should be covered for every connector (e.g. a rate limiting test case)?

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py [hunk 15]
```diff
@@ -197,6 +198,12 @@ def request_headers(self, **kwargs) -> Mapping[str, Any]:
             headers["Stripe-Account"] = self.account_id
         return headers
 
+    def retry_factor(self) -> float:
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:201]: There are two conflicting wishes here:
* Testing the retry mechanism with backoff
* Making sure tests run very fast

I decided to go with the tradeoff of not testing the fact that we sleep during backoff but still testing the retry mechanism so that tests can run fast

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 14 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:262]: This is something I was exploring. The issue I have is that from PyCharm, I can't test a specific method because of that which is annoying for the development loop. Also, these might not be relevant for some APIs because there is no 429/rate limiting which would be odd to force devs to implement it. In the end, I'm not sure it's worth. Do you have a small experiment that would help us confirm or infirm this?

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 11 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:51]: I agree with this. I'll make them separate

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 12 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:86]: Can you provide an example of "Events stream" changes and how both solutions would be affected? I would like to make sure I understand the scope of this comment

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 13 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:122]: The method name is definitely not enough. I'll update the test to add URL doc and explanation

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (clnoll) [reply airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:201]: Yeah that makes sense to me. Since backoff is its own library theoretically we shouldn't have to provide its test coverage.

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 12 above)
Comment (clnoll) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:86]: Hmm there are lots of ways I can think of doing this, but one idea might be to have a named tuple that represents a stream that's being tested and holds any information that we consider to define that stream. For instance if a stream is always going to have a particular URL, that would go into the named tuple, along with whatever query params are allowed, etc.
The other solution would just be to pull the URL out into a constant, and maybe list out, e.g., query params that we support for it somewhere along with a link to the documentation.

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 13 above)
Comment (clnoll) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:122]: WDYT about separating tests that feel specific to Stripe from tests that we expect to implement for all sources?

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 14 above)
Comment (clnoll) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:262]: I'm not sure I quite understand the pycharm issue. 

I think that for connectors who don't need rate limiting, for example, they could just override the abstractmethod with a dummy method that just does a `pass`. The nice thing about doing that is that it makes it obvious that that particular source does not need to be concerned with rate limiting (it's much harder to look for the absence of a test).

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 11 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:51]: I added the change in the CDK changes branch (see https://github.com/airbytehq/airbyte/pull/33305/commits/b5b4392b3101a730fe94789aec0365ebb9585752) and updated this branch

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 12 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:86]: A stream will have multiple URLs for stripe though as depending on the state provided, it'll use the resource associated to the stream (for example, "application_fess") or events. This is a special case for stripe as the event stream doesn't have such weird logic but it's almost the only one (this and accounts I think)

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 13 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:122]: Can you give me an example of tests that we expect all sources to implement vs source specific? I can be for another source than Stripe. I just want to see what we can generify. Right now, I'm leaning to "we can't generify anything" as even things I would take for granted like record seclection is not implemented by all the streams (if the API returns a list for example)

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 14 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:262]: Based on the information you added, I misunderstood your point. PyCharm should handle this just fine.

So what I like about this idea is that it's a nice checklist for the dev to make sure he tested all the base use cases. What I don't like about this is the maintenance: if we add a test method in `BaseFullRefreshTest` that is not relevant to my streams and that I have 46 streams like source-stripe has, I would need to update up to 46 tests classes without it providing value to my connector. I can see cases like this pop up somewhat often, one coming soon would be that sources emit state messages on full refresh (but it might not be a feature that makes sense to implement for connector X ASAP). 

So basically, I'm torn about this. I'm wondering if there would be another way to document those features without forcing tests on it. Please, challenge my concerns though because I don't feel 100% convinced about it

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (girarda) [reply airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:24]: @maxi297 [the platform sets an environment variable when it's running on cloud](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/utils/is_cloud_environment.py#L17-L18), but I agree with Catherine that cloud is not mutually exclusive with testing so two variables make sense

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (girarda) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:229]: I suspect this can't happen for events but we do skip some streams with the test account

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (girarda) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:241]: > I don't know if it's the case per se for stripe. I would assume not because right now, the failure type is system_error and not config_error.

can we file an issue and add it to the api-connectors team's backlog? we definitely don't want to be paged because 3 users failed to authenticate

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 14 above)
Comment (girarda) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:262]: > I would need to update up to 46 tests classes without it providing value to my connector

this does sound annoying, but I think seeing it as a checklist is right. connector developers should write tests or acknowledge they should've written one but explicitly skipped it for a given stream.

^feel free to challenge that statement, but I like the direction of enforced checklists. it seems more scalable than relying on a reviewer to verify a checklist has been followed

Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:262]: An issue has been created for this [here](https://github.com/airbytehq/airbyte/issues/33417)

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:225]: Found it! I'll update this as part of https://github.com/airbytehq/airbyte/pull/33352/ since I moved the payload to a common place

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:242]: Added [here](https://github.com/airbytehq/airbyte/issues/33420)

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 14 above)
Comment (clnoll) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:263]: I suppose developers could also make a `BaseStripeFullRefreshTest(BaseFullRefreshTest)` that indicates which ones aren't needed for Stripe, and they inherit from that for the 46 streams.

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (clnoll) [reply airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:201]: Actually - can we monkeypatch this in conftest instead of wrapping it in production code?

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 12 above)
Comment (clnoll) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:87]: Okay interesting. I still think that it would be nice to have something that serves as a centralized place to show exactly which URLs we're hitting, which query params we're supporting etc, even if it's just a bunch of constants and helper functions in a file. For example we're using a hardcoded URL and interpolating query param values in the same way in a bunch of places.
We should establish a pattern for this in the Stripe connector because it's going to be used as a sort of template for adding integration tests for other connectors.

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 13 above)
Comment (clnoll) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:123]: At a glance here are some that seem more generic to me, but tell me if I'm wrong:
- `test_given_lookback_window_when_read_then_request_before_start_date`
- `test_given_slice_range_when_read_then_perform_multiple_requests`
- `test_given_http_status_400_when_read_then_stream_is_ignored`
- `test_given_http_status_401_when_read_then_stream_is_incomplete` << not totally confident about this one
- `test_given_slice_range_when_read_then_perform_multiple_requests`
- `test_when_read_then_validate_availability_for_full_refresh_and_incremental`
... and lots more. Basically it feels like these could be abstractmethods that we could expect integration tests for each source to override.

Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:123]: It feels like a good basis for https://github.com/airbytehq/airbyte/issues/33417. I'll add these to the ticket

Comment (clnoll) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:123]: Thank you!

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 12 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py]: I did create a "StripeRequestBuilder" which would handle the transformation of type and the constants like the base URL. I think that it's at least adding readability but I would like you to confirm if this addresses your concern

Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py]: As discussed in another conversation, this does not really addresses the concern. The concern is that it would be nice for a dev to figure out the integration with the API with an object that describes those. A draft to illustrate this is:
```
class StripeEvents:
    full_refresh_url = "https://api.stripe.com/v1/events"
    incremental_url= "https://api.stripe.com/v1/events/incremental"  # or whatever this really is
    header_template = '{"Stripe-Account": "{}", "Authorization": "Bearer {}"}'
    query_params_template = '{"created[gte]": {}, "created[lte]": {}}'

    def build_query_params(self, gte, lte):
        self.query_params_template.format(gte, lte)

    def build_headers(self, account_id, token):
        self.header_template.format(account_id, token)
```

This helps the dev visualize that there are two different endpoints involved and which parameters it involves. There is a bit of complexity to it as something a sync mode can have multiple endpoints with different query params each. We will iterate on this idea while writing tests for other streams and see where it lands

**Review by girarda — APPROVED** [review]
:shipit:

File: airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:250]: It seems like the team tried to add the state in the availability strategy in order to check properly (see https://github.com/airbytehq/airbyte/pull/32223). It feels like this is a case where we instantiate things without all the information and hence we need to update the library which we agreed is not a good pattern. 

My feeling is that when we will have full_refresh with state, we will need to rethink that so I would punt on that

**Review by clnoll — APPROVED** [review]
